### PR TITLE
planner: add CloneForPlanCache method to Plan interface

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #54057

Problem Summary: planner: add `CloneForPlanCache` method to Plan interface

### What changed and how does it work?

planner: add `CloneForPlanCache` method to Plan interface

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_cache_rebuild_test.go [hunk 1]
```diff
@@ -29,38 +31,53 @@ func TestPlanCacheClone(t *testing.T) {
 	store := testkit.CreateMockStore(t)
 	tk := testkit.NewTestKit(t, store)
 	tk.MustExec(`use test`)
-	tk.MustExec(`create table t (a int, b int, c int, primary key(a), key(b))`)
+	tk.MustExec(`create table t (a int, b int, c int, d int, primary key(a), key(b), unique key(d))`)
+
+	for i := -20; i < 20; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values (%v,%v,%v, %v)", i, rand.Intn(20), rand.Intn(20), -i))
```
Comment (hi-rustin) [inline pkg/planner/core/plan_cache_rebuild_test.go:37]:
```suggestion
		tk.MustExec(fmt.Sprintf("insert into t values (%v,%v,%v,%v)", i, rand.Intn(20), rand.Intn(20), -i))
```

**Review by hi-rustin — APPROVED** [review]

File: pkg/planner/core/physical_plans.go [hunk 2]
```diff
@@ -736,7 +755,7 @@ type PhysicalIndexScan struct {
 	// The index scan may be on a partition.
 	physicalTableID int64
 
-	GenExprs map[model.TableItemID]expression.Expression
+	genExprs map[model.TableItemID]expression.Expression
```
Comment (AilinKid) [inline pkg/planner/core/physical_plans.go:758]: why make it for lower case？

in a later pkg move, the physical operator will be moved out of the core, and elements inside should exported for usage.

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Pull request merged**